### PR TITLE
keycloak: 16.1.0 -> 17.0.1

### PR DESCRIPTION
**Pull request by talyz**
###### Description of changes
Update Keycloak to the latest version.

We have to use the `-legacy` version, since the module has to be ported from Wildfly to Quarkus before we can switch. The switch has to be made [before June](https://www.keycloak.org/migration/migrating-to-quarkus).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/keycloak/default.nix [hunk 1]
```diff
@@ -19,11 +19,11 @@ let
 in
 stdenv.mkDerivation rec {
   pname   = "keycloak";
-  version = "16.1.0";
+  version = "17.0.1";
```
Comment (ngerstle) [inline pkgs/servers/keycloak/default.nix:22]:
```suggestion
  version = "17.0.1-legacy";
```
How do we integrate the non-legacy/regular 17.0.1 version, if we implicitly link keycloak-17.0.1 to the legacy package?

Comment (talyz) [reply pkgs/servers/keycloak/default.nix:22]: Well, I figured we'd just do a hard switch-over - rewrite the module and switch the package at the same time. Unless there are users who use the package without the module, this should work fine. If there are, and they need the legacy package, i guess we need to package both.

Comment (ngerstle) [reply pkgs/servers/keycloak/default.nix:22]: hmm- I'd prefer a separate version to indicate legacy, and follow the upstream versioning scheme in the version here.
We could add a check to the module to require <17 || 17*-pre, at least until we're able to update the module (and then assert >= 17)..

Comment (ngerstle) [reply pkgs/servers/keycloak/default.nix:22]: Given https://bugzilla.redhat.com/show_bug.cgi?id=2013577, I can live with either versioning option, however.

Comment (ngerstle) [reply pkgs/servers/keycloak/default.nix:22]: (I'm not sure if 
```suggestion
  version = "17.0.1.legacy";
```
would be better)

**Comment by ngerstle** [comment]
I'm working on the update to quarkus in https://github.com/NixOS/nixpkgs/pull/167036 - still need to check out the plugins process, the nixos module, and tests..

**Pull request merged**